docs/src/appendix/experimental-features.md

```diff
@@ -143,56 +143,156 @@ Consider a Interval with a binary point of 3: aaa.bbb
 | shiftLeftBinaryPoint(2) | a.aabbb |  5 | X | X  | increase the precision |
 | shiftRighBinaryPoint(2) | aaaa.b |  1 | X | X  | reduce the precision |
 
-## Loading Memories in simulation <a name="loading-memories"></a>
+## Loading Memories for simulation or FPGA initialization <a name="loading-memories"></a>
 
-Chisel now supports an experimental method for annotating memories to be loaded from a text file containing
-hex or binary numbers. When using verilog simulation it uses the `$readmemh` or `$readmemb`
-verilog extension. The treadle simulator can also load memories using the same annotation.
+Chisel supports multiple experimental methods for annotating memories to be loaded from a text file containing hex or binary data. When using verilog simulation it uses the `$readmemh` or `$readmemb` verilog extension. The treadle simulator can also load memories using the same annotation.
 
-### How to annotate
-Assuming you have a memory in a Module
-```scala mdoc:invisible
+### Inline initialization with external file
+
+Memories can be initialized by generating inline `readmemh` or `readmemb` statements in the output Verilog.
+
+The function `loadMemoryFromFileInline` from `chisel3.util.experimental` allows the memory to be initialized by the synthesis software from the specified file. Chisel does not validate the file contents nor its location. Both the memory initialization file and the Verilog source should be accessible for the toolchain.
+
+```scala mdoc:silent
 import chisel3._
-val memoryDepth = 8
-val memoryType = Bool()
-```
-```scala
-val memory = Mem(memoryDepth, memoryType)
+import chisel3.util.experimental.loadMemoryFromFileInline
+
+class InitMemInline(memoryFile: String = "") extends Module {
+  val width: Int = 32
+  val io = IO(new Bundle {
+    val enable = Input(Bool())
+    val write = Input(Bool())
+    val addr = Input(UInt(10.W))
+    val dataIn = Input(UInt(width.W))
+    val dataOut = Output(UInt(width.W))
+  })
+
+  val mem = SyncReadMem(1024, UInt(width.W))
+  // Initialize memory
+  if (memoryFile.trim().nonEmpty) {
+    loadMemoryFromFileInline(mem, memoryFile)
+  }
+  io.dataOut := DontCare
+  when(io.enable) {
+    val rdwrPort = mem(io.addr)
+    when (io.write) { rdwrPort := io.dataIn }
+      .otherwise    { io.dataOut := rdwrPort }
+  }
+}
 ```
 
-At the top of your file just add the import 
+The default is to use `$readmemh` (which assumes all numbers in the file are in ascii hex),
+but to use ascii binary there is an optional `hexOrBinary` argument which can be set to `MemoryLoadFileType.Hex` or `MemoryLoadFileType.Binary`. You will need to add an additional import.
+
+By default, the inline initialization will generate the memory `readmem` statements inside an `ifndef SYNTHESIS` block, which suits ASIC workflow, or outside the block suiting FPGA workflows.
```

Comment:
```suggestion
By default, the inline initialization will generate the memory `readmem` statements inside an `ifndef SYNTHESIS` block, which suits ASIC workflows.
```

Right? For FPGA workflows you have to do all the stuff below? (Again I don't know the right answer, but this sentence seems to contradict the below).

Reply:
Totally makes sense! Thanks :)